# archaic-chrX-sexbias
Code and data to reproduce main figures and simulations of Chevy, Huerta-Sanchez, and Ramachandran, 202X.

## Files
### `plotting`
  - `figX_plot.py` contains all the code necessary to create Fig X.
  - `figX_data/` contains all data necessary to plot Fig X.  Datafiles from simulations are provided, and were generated using the scripts provided in `generating_simulation_data/`.  Sources for previously published data are provided in `README.md` files. 
  - `fig4_prep.py` contains the code for pickling the data necessary for plotting each panel of Fig 4 into four files: `fig4_tracts_[ABCD].pkl`.  These `.pkl` files are provided. 
### `generating_simulation_data`
 - `sim_seq_info.txt` contains the exon coordinates and recombination rate map used in SLiM simulations.
 - `simulation.slim` is an example SLiM simulation script used to generate simulated archaic coverage data.  In order to run a simulation, confirm that the:
   - paths to `sim_seq_info.txt` (ln. 17) and the output treesequence `.trees.decap` (ln. 106) are set correctly,
   - the inheritance method `"A"` for autosomal or `"X"` for chromosome X (ln. 4) is correct,
   - the male fraction of the archaic introgressors (ln. 5) is correct,
   - the dominance coefficient $h$ (second argument to `initializeMutationType`, ln. 12) is correct, and
   - the "timecourse" section (ln.s 96-99) is uncommented if you want to calculate archaic coverage in samples during the generations immediately following introgression.
 - `tree_tools.py` contains helper functions used in `ibd_to_pI.py` and `ibd_to_spectra.py`.
 It is imported as a module, so must be in the same directory or added to the path using `sys.path.append('/path/to/tree_tools.py')`.
 - `ibd_to_pI.py` calculates the archaic coverage on haplotypes sampled at the end of a simulation.  
 It takes a treesequence file (`.trees.decap`) generated by `simulation.slim` and generates two further files:  
    - a tab-delimited text file (`.pI`) where the first column is the sample ID of the haplotype and the second column is the archaic coverage on that sample at the end of the simulation, and, 
    - optionally, a pickled copy (`.ibd`) of a dictionary containing the start and end coordinates of each archaic coverage tract within each sample.
    This is only generated if the `save_ibd_dicts` variable in `main()` is set to `True`.
    Generating the `.ibd` file is necessary to calculate the coverage tract length spectra using `ibd_to_spectra.py`.
 - `ibd_to_spectra.py` calculates the lengths of archaic coverage tracts found on sampled haplotypes.  It takes an `.ibd` file generated by `ibd_to_pI.py` and generates a compressed version (`.tracts.npz`) of the tract length spectra across all samples, specific to each human sample, and specific to each archaic introgressor.  


## Citation
TBA
